sqlx-postgres/src/migrate.rs

```diff
@@ -208,33 +208,49 @@ CREATE TABLE IF NOT EXISTS _sqlx_migrations (
         Box::pin(async move {
-            let mut tx = self.begin().await?;
             let start = Instant::now();
-
-            // Use a single transaction for the actual migration script and the essential bookeeping so we never
-            // execute migrations twice. See https://github.com/launchbadge/sqlx/issues/1966.
-            // The `execution_time` however can only be measured for the whole transaction. This value _only_ exists for
-            // data lineage and debugging reasons, so it is not super important if it is lost. So we initialize it to -1
-            // and update it once the actual transaction completed.
-            let _ = tx
-                .execute(&*migration.sql)
-                .await
-                .map_err(|e| MigrateError::ExecuteMigration(e, migration.version))?;
-
-            // language=SQL
-            let _ = query(
-                r#"
+            if !migration.no_tx {
+                let mut tx = self.begin().await?;
+
+                // Use a single transaction for the actual migration script and the essential bookeeping so we never
+                // execute migrations twice. See https://github.com/launchbadge/sqlx/issues/1966.
+                // The `execution_time` however can only be measured for the whole transaction. This value _only_ exists for
+                // data lineage and debugging reasons, so it is not super important if it is lost. So we initialize it to -1
+                // and update it once the actual transaction completed.
+                let _ = tx
+                    .execute(&*migration.sql)
+                    .await
+                    .map_err(|e| MigrateError::ExecuteMigration(e, migration.version))?;
+
+                // language=SQL
+                let _ = query(
+                    r#"
     INSERT INTO _sqlx_migrations ( version, description, success, checksum, execution_time )
     VALUES ( $1, $2, TRUE, $3, -1 )
                 "#,
-            )
-            .bind(migration.version)
-            .bind(&*migration.description)
-            .bind(&*migration.checksum)
-            .execute(&mut *tx)
-            .await?;
+                )
+                .bind(migration.version)
+                .bind(&*migration.description)
+                .bind(&*migration.checksum)
+                .execute(&mut *tx)
+                .await?;
 
-            tx.commit().await?;
+                tx.commit().await?;
+            } else {
+                query(&migration.sql).execute(&mut *self).await?;
+                // language=SQL
+                let _ = query(
+                    r#"
+    INSERT INTO _sqlx_migrations ( version, description, success, checksum, execution_time )
+    VALUES ( $1, $2, TRUE, $3, -1 )
+                "#,
+                )
+                .bind(migration.version)
+                .bind(&migration.description)
+                .bind(&*migration.checksum)
+                .execute(&mut *self)
+                .await?;
```

Comment:
You could deduplicate this by putting it in a function that takes `&mut PgConnection` since `Transaction` derefs to it.

Reply:
Yes, great suggestion!